src/system_simulation/simulation_runner.rb

```diff
@@ -1,111 +1,160 @@
-    #AnimationDataSample.new(0.0, mocked_position_data)
+    raw_data.index(equilibrium_data_row)
   end
 
+  def constant_for_constrained_angle(allowed_angle_delta = Math::PI / 2.0, initial_constant = 500, mass = 20,
+                                     spring_id = 0, angle_id = 0)
+    # steps which the algorithm uses to approximate the valid spring constant
+    step_sizes = [1500, 1000, 200, 50, 5]
+    constant = initial_constant
+    step_size = step_sizes.shift
+    keep_searching = true
+    abort_threshold = 50_000
+    while keep_searching
+      # puts "Current k: #{constant} Step size: #{step_size}"
+      run_simulation(constant, mass, 'revLeft.phi')
+      if !angle_valid(read_csv, allowed_angle_delta)
+        # increase spring constant to decrease angle delta
+        constant += step_size
+      elsif !step_sizes.empty?
+        # go back last step_size
+        constant -= step_size
+        # reduce step size and continue
+        step_size = step_sizes.shift
+      else
+        # we reached smallest step size and found a valid spring constant, so we're done
+        keep_searching = false
+      end
+
+      keep_searching = false if constant >= abort_threshold
+    end
+
+    constant
+  end
 
   private
 
-  def run_compilation()
-    output, signal = Open3.capture2e("cp #{@model_name}.mo  #{@directory}", :chdir => File.dirname(__FILE__))
+  def angle_valid(data, max_allowed_delta = Math::PI / 2.0)
+    data = data.map { |data_sample| data_sample[1].to_f }
+    # remove initial data point
+    data.shift
+
+    delta = data.max - data.min
+    puts "delta: #{delta} maxdelta: #{max_allowed_delta} max: #{data.max}, min: #{data.min}, "
+    delta < max_allowed_delta
+  end
+
+  def run_compilation
+    output, _ = Open3.capture2e("cp #{@model_name}.mo  #{@directory}", chdir: File.dirname(__FILE__))
     p output
-    output, signal = Open3.capture2e("omc -s #{@model_name}.mo && mv #{@model_name}.makefile Makefile && make -j 8", :chdir => @directory)
+    output, _ = Open3.capture2e("omc -s #{@model_name}.mo && mv #{@model_name}.makefile Makefile && make -j 8",
+                                chdir: @directory)
     p output
   end
 
-  def run_simulation(constant, mass, filter="*")
+  def run_simulation(constant, mass, filter = '*')
     # TODO adjust sampling rate dynamically
-    overrides = "outputFormat='csv',variableFilter='#{filter}',startTime=0.3,stopTime=10,stepSize=0.02,springDamperParallel1.c='#{constant}'"
+    overrides = "outputFormat='csv',variableFilter='#{filter}',startTime=0.3,stopTime=10,stepSize=0.1,springDamperParallel1.c='#{constant}'"
     command = "./#{@model_name} -override #{overrides}"
     puts(command)
-    Open3.popen2e(command, :chdir => @directory) do |i, o, t|
-      o.each {|l| puts l }
-      status = t.value
+    Open3.popen2e(command, chdir: @directory) do |i, o, t|
+      o.each { |l| puts l }
```

Comment:
Maybe just add a comment what this does

Reply:
Done in 6ee258a3c58b1fd820677f52efe4907d283251b1